src/terms.ml

```diff
@@ -203,6 +205,50 @@ type term =
     can hence be substituted in [r.rhs] with [Bindlib.msubst r.rhs env] to get
     the result of the application of the rule. *)
 
+(** {3 Decision trees for rewriting} *)
+
+(** Trees are used to efficiently choose a rewriting rule given a list of
+    terms (beginning with a symbol) to be rewrote.  The left hand has side of
+    the rule is spread across the {!constructor:Node}s of the tree.  Hence,
+    progressing down the tree is equivalent to reducing the set of possible
+    rules.  When a {!constructor:Leaf} is reached, the target is rewrote to
+    the content of the leaf. *)
+ and tree =
+  | Leaf of int IntMap.t * (term_env, term) Bindlib.mbinder
+  (** Holds the right hand side of a rule.  In a {!constructor:Leaf}[(e, a)],
+      - [a] is the right hand side of the rule.
+      - [e] maps a position in the stack containing terms which stand as
```

Comment:
Why not using an `int array`?

Reply:
I have to check, but the set of keys may not be an interval of integers.

Reply:
I see: not all terms of the stack correspond to LHS variables, right?

Reply:
But, for a given RHS [a], couldn't we have an array [e] of length [Bindlib.mbinder_arity a] so that [e.(i)] gives the position in the stack of the term corresponding to the [i]-th variable?

Reply:
It's not that not all term of the stack are LHS variables, but there might be--in the stack--terms that matched variables for other rules (that were potentially applicable before)

Reply:
In any case, we have to traverse the `env_builder`, so `list` or `array`, it doesn't really matter.  I've changed it to an associative list